src/librustc_mir/build/expr/into.rs

```diff
@@ -169,39 +169,46 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
 
-                let might_break = this.in_loop_scope(loop_block, exit_block, move |this| {
-                    // conduct the test, if necessary
-                    let body_block;
-                    if let Some(cond_expr) = opt_cond_expr {
-                        // This loop has a condition, ergo its exit_block is reachable.
-                        this.find_loop_scope(expr_span, None).might_break = true;
+                let has_cond = opt_cond_expr.is_some();
+                // `loop` loops admit an argument to `break`
+                let break_destination = opt_cond_expr.is_none();
 
-                        let loop_block_end;
-                        let cond = unpack!(loop_block_end = this.as_operand(loop_block, cond_expr));
-                        body_block = this.cfg.start_new_block();
-                        this.cfg.terminate(loop_block_end, source_info,
-                                           TerminatorKind::If {
-                                               cond: cond,
-                                               targets: (body_block, exit_block)
-                                           });
+                // conduct the test, if necessary
+                let body_block;
+                if let Some(cond_expr) = opt_cond_expr {
+                    let loop_block_end;
+                    let cond = unpack!(loop_block_end = this.as_operand(loop_block, cond_expr));
+                    body_block = this.cfg.start_new_block();
+                    this.cfg.terminate(loop_block_end, source_info,
+                                       TerminatorKind::If {
+                                           cond: cond,
+                                           targets: (body_block, exit_block)
+                                       });
+                } else {
+                    body_block = loop_block;
+                }
+
+                let might_break = this.in_loop_scope(
+                    loop_block, exit_block,
+                    if break_destination {
+                        Some(destination.clone())
                     } else {
-                        body_block = loop_block;
+                        None
+                    },
+                    move |this| {
+                        // The “return” value of the loop body must always be an unit. We therefore
+                        // introduce a unit temporary as the destination for the loop body.
+                        let tmp = this.get_unit_temp();
+                        // Execute the body, branching back to the test.
+                        let body_block_end = unpack!(this.into(&tmp, body_block, body));
+                        this.cfg.terminate(body_block_end, source_info,
+                                           TerminatorKind::Goto { target: loop_block });
                     }
-
-                    // The “return” value of the loop body must always be an unit, but we cannot
```

Comment:
AFAICT the comment is about reusing the value of the loop's block (which is always `()`) as the value of the loop itself. When you consider that `break EXPR` exists, though, it becomes apparent (IMO) that this is nonsensical because the value of the loop is not always `()`. That's why I removed most of the comment.